### ISSUE TITLE
compute coads only once for initialisation

### ISSUE THREAD

**Issue by herjy** [pmelchior/scarlet#172]
Sources are initialised by building coadds weighted by a pre-computed sed and applying a set of morphological constraints on the coadds to get a morphology. This means that the coad has to be computed for each sed, resulting in computing several time the same multiplication weigthed differently depending on background rmses. This is generaly cheap, but suboptimal. On top of this, the weighting ensures that the amplitude of the morphologies correspond to the amplitude of the seds. However, the morphologies are now normalised to their max at unity so knowing this we could just create one coadd weighted by the background levels and initialise all sources on that. The current initialisation is robust to a general implementation where sources are not max-normalised, but do we really need to support this? 

Also the current architecture requires that we pass raw observations to the function that build coads. In the multiresolution case, this would lead to either:
* performing an interpolation of all low-res images for each source (highly suboptimal!) or
* having to create synthetic `Observation` for each `LowResObservation` that contains the interpolated images at high-res. Possible, but I find it inelegant.

I propose to refactor the initialisation and have the coadd created only once. The morphologies would be initialised with a max normalisation on top of the rest of the current initialisation.

**Comment by herjy** [comment]
I think the best way to do this is to make a rather big overhaul of the initialisation. I'm thinking of an initialisation class that would take coordinates, frames, observations and a list of source classes to perform the initialisation. This would be, I think, more user-friendly than the weird for loop with `if` statements to initialise each source one by one.

**Comment by fred3m** [comment]
I'm not sure that I agree with your interpretation, but I might be missing something. When we build the detection coadd we weight all bands with the SED of each source not because we want the amplitude to be correct (since we normalize it away when we factor in the SED and PSF) but because sources have power in different parts of the spectrum. So a faint red source requires a different coadd than a bright blue source, where the faint source needs more weight given to the redder bands to properly initialize its morphology (and sometimes to even be detected at all).

With that being said, I've been looking at the detection coadds lately for sources of different colors and they aren't _that_ different. So it might be that for most sources using a common coadd will save enough time to be worth the slight degradation in initialization, whereas other times it will be inappropriate. So my recommendation would be to create a `detection_coadd` parameter that can be passed to a source. If `detection_coadd` is `None` then the coadd will be rebuilt for a given source, otherwise it will use the joint `detection_coadd`. Then it will be up to our custom initialization schemes (which will be different for different use cases) to decide which sources can use a joint coadd and which ones will need their own custom coadd.

**Comment by fred3m** [comment]
To address you other comment, I do agree with an overhaul of the way that we do initialization but in a different way. The way scarlet 0.5 was designed was to make the `Component` class the core object and all of the source classes were just the `Component` with different initializations and constraints that could be defined by the user.

This changed in scarlet 1.0 but didn't really go far enough. Now `Component` is subclassed based on different internal models, and the different source classes now differ in more than just the initialization. So `PointSource`, `ExtendedSource`, `MultiComponentSource`, `StarletSource`, et.c are really different models now, with different internal structures but reusing many of the same elements for initialization. But in some cases, as you mentioned, we might want the initialization to be different while still maintaining the same internal structure.

So for this I propose allowing the user to pass an initialization method to `Source.__init__`, which will make it easier to slightly change initializations without requiring a ton of parameters. The latest version of `init_extended_source` that Peter wrote does a good job breaking the initialization into modular bits and this should probably remain the default initialization, and makes it easier for users to use it as a model to write their own custom initialization methods.

Similarly, I think that we should make `constraints` a parameter that users can pass to `Source.__init__` again and remove the `monotonic` and `symmetric` arguments. My thinking is that if the user doesn't have a custom list of `constraints` then we just fall back to the default, which uses monotonicity and not symmetry. We can still build the `ConstraintChain` internally, unless `constraints` is a `ConstraintChain`.

I feel like this model makes our existing source classes more usable and gives the right amount of control to users.

**Comment by herjy** [comment]
So, on your about SEDs I agree that multipllying by a source's SED is like multiplying this source in the field by a collinear vector (I'm making shortcuts, you already have the idea). The trouble is that I'm not sure it helps faint sources. Since the SED is estimated on a single pixel at the  detection position, the SED of a faint source is a very noisy measurement and can even be erroneous if close to a bright source with a different colour, which I find a lot especially when I inform the detection with higher resolution data. So I'm not sure this helps faint source a lot and bright sources won't care anyway. 

Currently, both initialisations are still possible. I implemented a new initialise function that reduces the initialisation to this:
```sources = initialise(frame, obs, [(ra, dec) for ra, dec in zip(ras, decs)], scarlet.Extended(symmetric = False))
blend_multi = scarlet.Blend(multi_sources, obs)```
Here all sources are treated as  ExtendedSource. The last argument can be replaced by a list of Sources that can take optional parameters for the actual `ExtendedSource` (or other source). 
But the old initialisation is still possible, so that the current API is maintained. I'll have to make a few tweaks to ensure that the right coadd is used if we're settled on this, but your approach is easily feasible. Anyway, this implementation is beautiful, now there are no notions of `LowResObservation` or `obs_idx` on the user side. Handling these is completely transparent from the point of view of the user.

**Comment by herjy** [comment]
Yes, I'm having trouble with the various different Source classes that are multiplying. I think we should find a more flexible way to do that. Maybe the kind of source we want to use should be a keyword argument. I think this is outside of the scope of what I'm proposing at the moment, but I'd be happy to take that on. It would also be rather easy to do in my new initialisation scheme.

**Comment by fred3m** [comment]
I agree with you in general about SED weighted coadds. I just looked at a number of different blends and plotted the two detection coadds that had the largest total difference `sum(abs(coadd1-coadd2))` and the largest single pixel difference `max(abs(coadd1-coadd2))` and for the majority of blends the difference was barely recognizable. But for faint sources it does seem to make a difference:

Full color image:
![image](https://user-images.githubusercontent.com/2251900/82216978-bf259500-98e7-11ea-8b29-fb95fdfe9a6c.png)

Detection coadds:
![image](https://user-images.githubusercontent.com/2251900/82217014-cd73b100-98e7-11ea-887d-6de2a47ada4e.png)

The left image shows the detection coadd for the source with the blue x while the right shows the coadd for the source with the red x. So in the redder coadd the blue source doesn't have enough flux to even be initialized because it has zero (or negative) flux at the peak location.

But in all reality something with SNR that low should probably just be initialized with the PSF (even if it is an `ExtendedSource`) so you are probably right that in general we can use the weighted detection coadd for all sources (which would probably save a decent amount of time for fields with dozens to hundreds of sources.

**Comment by fred3m** [comment]
I address your second comment in #171 .

**Comment by herjy** [comment]
Thanks a lot for the study! If you look at image 4, and even 2 itmakes a difference though!
I like the idea of the detection informing the choice of sources, but I'm afraid that it would cause problems for faint, diffuse galaxies.

**Comment by fred3m** [comment]
True, but interestingly (for the reasons you mentioned in relation to the noise on the fainter sources) sources 2 and 4 didn't look as distinct in their own detection coadds as they do for source 5. This is why an earlier version of the code actually just used the band that a source was brightest in for detection without building a coadd at all.

**Comment by pmelchior** [comment]
I have some concerns about the source initialization, too (in that I think that the SEDs could be retuned *after* we initialized all sources, but that's an aside). Here's the core of the proposal:

```
sources = initialise(frame, obs, [(ra, dec) for ra, dec in zip(ras, decs)], 
                     scarlet.Extended(symmetric = False))
```
@herjy: you characterized this as transparent. It is much more concise but much less transparent or flexible. It assumes that all sources are `ExtendedSource`. If that's not the case, what's the logic of picking source types? What about users wanting to declare their own sources? How about a better star-galaxy separator? In #171, Fred summarized correctly that we should not specify *what* the user wants to fit, but give our best effort to implement useful code for *how* to fit it.

I believe that your proposal is driven by the wish to reuse the detection coadd, which is a good idea. I also agree with the concern that faint sources will have an unreliable peak SED, so that the initial morphology might be even worse than with flat or inverse variance weights. The main reason for using the SED is twofold:
* improved detection SNR
* reduced impact of neighbors (with different SED) for morphology initialization 

So, how do we make joint initialization happen? The easiest option is @fred3m's proposal of providing a custom `detection_coadd` for every source. How to construct that is left to the user (for instance LSST might construct its own detection coadd). I would also suggest to provide a reference implementation in the multi-resolution case as part of `scarlet` because this is about *how* not about *what*.

**Comment by herjy** [comment]
As mentioned above, the last argument can be replaced by a list of sources. For instance:
```source_list = [Point(), MultiComponent(), *([Extended()]*(len(catalog)-2))]
sources = initialise(model_frame, 
                     observation, 
                     [(r, d) for r, d in zip(catalog['y'], catalog['x'])],
                     source_list)```
So users can specify whatever sources they want. 
So far the previous intialisation scheme (which computes coadds) is still supported and even benefits from and extra set of optional parameters that allow the users to provide a `coadd` and `bg_cutoff`, so we don't lose any flexibility.

I completely get the advantages of the SED coadd. To be fair, my method it is motivated by runtime because multi-resolution coadds are not cheap. Though the arguments in favour of the SED fall short for faint sources for the reasons mentioned above. I don't like that the initialisation be up to the user. because right now I'm the user, and automating initialisation makes my scripts cleaner. 
1. because right now I'm the user, and automating initialisation makes my scripts cleaner.
2. because the initialisation is extremely delicate and I think users will have a much better experience if we have clean basic initialisation codes. If they have something better, sure, they can still provide it.

**Comment by herjy** [comment]
resolved as part of PR #173
